8271949: dumppath in -XX:FlightRecorderOptions does not affect

YaSuenag opened the pull request:
We set `dumppath` in `-XX:FlightRecorderOptions` or `JFR.configure dumppath` jcmd when we want to set emergency dump path. However they do not afffect.

### -XX:FlightRecorderOptions

```
$ java -XX:FlightRecorderOptions=dumppath=`pwd` -XX:StartFlightRecording -Xlog:exceptions=info --version
[0.032s][info][exceptions] Exception <a 'java/lang/IllegalArgumentException'{0x00000007470163b8}: Unknown argument 'dumppath' in diagnostic command.> (0x00000007470163b8)
thrown [open/src/hotspot/share/services/diagnosticFramework.cpp, line 215]
for thread 0x00007f4f700236a0
[0.033s][error][arguments ] Unknown argument 'dumppath' in diagnostic command.
Error occurred during initialization of VM
Failure when starting JFR on_create_vm_2
```

### jcmd

`jcmd` shows the configuration change was succeeded, but it would not affect.

```
$ jcmd 1046 JFR.configure dumppath=/tmp
1046:
Dump path: /tmp
```

YaSuenag commented:
@egahlin Can you review this PR?

YaSuenag commented:
PING: Could you review this PR? I think this fix is useful for containerized apps because it should store emergency dump to persistent storage. Now we have no chance to specify dump path due to this bug.

D-D-H commented on src/jdk.jfr/share/classes/jdk/jfr/internal/Options.java:
```diff
@@ -48,7 +48,7 @@
     private static final boolean DEFAULT_SAMPLE_THREADS = true;
     private static final long DEFAULT_MAX_CHUNK_SIZE = 12 * 1024 * 1024;
-    private static final SafePath DEFAULT_DUMP_PATH = SecuritySupport.USER_HOME;
+    private static final SafePath DEFAULT_DUMP_PATH = new SafePath(".");
```
Why do need to change DEFAULT_DUMP_PATH here?

YaSuenag replied:
Emergency dump will be dumped on current work directory now. This default value does not effect. So I changed it to current behavior, and it is nature because other dumps (hs_err, cireplay) are also dumped on current directory.

D-D-H replied:
Make sense, I think we could remove `SecuritySupport.USER_HOME` since no place uses it now.

YaSuenag replied:
I removed `SecuritySupport.USER_HOME` in new commit.

D-D-H reviewed (approved):
Looks good to me, but I am not a Reviewer:)

YaSuenag commented:
TestStringDeduplicationAgeThreshold was failed on Linux x64 on GitHub Actions, but it does not seem to be caused by this change. I'm waiting on Reviewer...

egahlin commented:
It would good to have tests. There could be platform specific issues and and we missed outages in the past.

YaSuenag commented:
@egahlin Thanks for your comment! I extended TestDumpOnCrash to check dumppath option. Could you review?

mgronlun commented:
Hi Yasumasa,

I noted the following when stepping through the patch:

1. Mismatch in default behaviour for paths between option "repository" and "dumppath" for relative paths:

-XX:FlightRecorderOptions=repository=apa,dumppath=ko

Repository path: $PWD/apa
Emergency dump path: ko

2. Modified default output in hs_err<pid>.log file:

From:
<pre>
# CreateCoredumpOnCrash turned off, no core file dumped"
#
# JFR recording file will be written. Location: D:\utilities\jtreg\runtime_artifacts\work\jdk\jfr\jvm\TestDumpOnCrash\hs_err_pid23516.jfr
#
# An error report file with more information is saved as:
# D:\utilities\jtreg\runtime_artifacts\work\jdk\jfr\jvm\TestDumpOnCrash\hs_err_pid23516.log
#
</pre>

To:
<pre>
# CreateCoredumpOnCrash turned off, no core file dumped
#
# JFR recording file will be written. Location: .\hs_err_pid23516.jfr <<--- not printing the full path as before
#
# An error report file with more information is saved as:
# D:\utilities\jtreg\runtime_artifacts\work\jdk\jfr\jvm\TestDumpOnCrash\hs_err_pid23516.log
#
</pre>

3. "Path to dump" description is vague. It is not clear that it must be directory, let alone that the directory must exist, and that you need write-access to it:

-XX:FlightRecorderOptions=maxchunksize=1m,repository=apa,dumppath=C:\temp\mydump.jfr ?

4. Lack of troubleshooting information:

Any problematic path, for example specifying mounts that do not exist, wrong spelling, specifying a file path instead of a directory path, lack of security etc will get this:
<pre>
#
# CreateCoredumpOnCrash turned off, no core file dumped
#
# The JFR repository may contain useful JFR files. Location: C:\Users\mgronlun\AppData\Local\Temp\2021_09_03_16_36_14_7524
#
# An error report file with more information is saved as:
# D:\utilities\jtreg\runtime_artifacts\work\jdk\jfr\jvm\TestDumpOnCrash\hs_err_pid7524.log
</pre>

Why was the not my stated dumppath=<path> taken into account and used in creating a emergency dump file?

Thanks
Markus

YaSuenag commented:
@mgronlun Thanks for your comment!

> 1. Modified default output in hs_err_pid.log file:

I fixed it.

> 1. Lack of troubleshooting information:

I checked dump path and write permission of it. However they might be changed when the dump happens, so I think it should be checked again in jfrEmergencyDump.cpp .
In hs_err log, it would fallback to current directory or temp directory if it cannot be written to specified path ( -XX:ErrorFile). Should emergency dump do so?

> 1. "Path to dump" description is vague. It is not clear that it must be directory, let alone that the directory must exist, and that you need write-access to it:

I guess you mentioned to the description in `JfrConfigureFlightRecorderDCmd::_dump_path`. As you know it is one-liner help message, so I think it is prefer to simple. In addition, I don't change semantics of `dumppath` (and don't change this description).

> 1. Mismatch in default behaviour for paths between option "repository" and "dumppath" for relative paths:

I cannot understand it well, but I guess you pointed "current directory" might be change since startup (start recording). But the same problem exists in current implementation because emergency dump path would be created by `os::get_current_directory()`, and it will be called when emergency dump happens.

YaSuenag commented:
@egahlin @mgronlun Could you review this PR? Please tell me if you have any comments.

mgronlun commented:
"I guess you mentioned to the description in JfrConfigureFlightRecorderDCmd::_dump_path. As you know it is one-liner help message, so I think it is prefer to simple. In addition, I don't change semantics of dumppath (and don't change this description)."

I do think the semantics have changed. From invariantly writing the emergency dump to the $PWD, to instead attempting to do it at a user-provided location, which comes with a new set of challenges.

"I cannot understand it well, but I guess you pointed "current directory" might be change since startup (start recording). But the same problem exists in current implementation because emergency dump path would be created by os::get_current_directory(), and it will be called when emergency dump happens."

I do not understand this comment. My point about a mismatch is that options "repository" and "dumppath" behave differently when you specify a relative path. "repository" starts out from the $PWD, "dumppath" does not.

YaSuenag commented:
> "I guess you mentioned to the description in JfrConfigureFlightRecorderDCmd::_dump_path. As you know it is one-liner help message, so I think it is prefer to simple. In addition, I don't change semantics of dumppath (and don't change this description)."
> 
> I do think the semantics have changed. From invariantly writing the emergency dump to the $PWD, to instead attempting to do it at a user-provided location, which comes with a new set of challenges.

I believe `dumppath` is for specifying the path to dump - @egahlin says so too in PR #4724 . In that context, this PR provides expected behavior in originally.

> "I cannot understand it well, but I guess you pointed "current directory" might be change since startup (start recording). But the same problem exists in current implementation because emergency dump path would be created by os::get_current_directory(), and it will be called when emergency dump happens."
> 
> I do not understand this comment. My point about a mismatch is that options "repository" and "dumppath" behave differently when you specify a relative path. "repository" starts out from the $PWD, "dumppath" does not.

No, the user can set `dumppath` with a relative path from `$PWD`.
You can see it in the change of jfrEmergencyDump.cpp . `get_dump_directory()` would build dump path without any other paths when the path is not specified.
In addition, I changed `Options::DEFAULT_DUMP_PATH` to `.` in this PR, then it works fine. It is the evidence that JFR can handle relative path for `dumppath` rightly.

mgronlun commented:
"No, the user can set dumppath with a relative path from $PWD.
You can see it in the change of jfrEmergencyDump.cpp . get_dump_directory() would build dump path without any other paths when the path is not specified.
In addition, I changed Options::DEFAULT_DUMP_PATH to . in this PR, then it works fine. It is the evidence that JFR can handle relative path for dumppath rightly."

Ok, you fixed it in the incremental changeset 'Set absolute path to JFR emergency dump path' - that's good.

I still see the modified output in the default case, i.e. when no dumppath is specified:

# CreateCoredumpOnCrash turned off, no core file dumped
#
# JFR recording file will be written. Location: D:\temp\SPECjbb2005\SPECjbb2005\.\hs_err_pid28564.jfr <<-- note the dot

"Check write permission on JFR emergency dump path in configuration" introduced an IllegalArgumentException if the dumppath= is invalid for some reason. This will lead to the JVM not starting:

[19.350s][error][jfr,startup] Cannot write JFR emergency dump to bad_location
java.lang.IllegalArgumentException: Cannot write JFR emergency dump to bad_location
Error occurred during initialization of VM
Failure when starting JFR on_create_vm_2

This can cause problems for users that pinpoint a location on a mounted drive, in case the mount is not reachable, perhaps because of a temporary network issue. In that case, their entire fleet of JVMs will not start. This might be too strong considering this is an option specifying what to do in an exceptional situation.

YaSuenag commented:
@mgronlun Thanks for the comment!
I fixed path resolution, and print warning message instead of throwing exception in new commit. Could you review again?

YaSuenag commented:
@egahlin @mgronlun Can you review this PR?

YaSuenag commented:
Ping!

mgronlun commented:
So how do you handle the case when it cannot be written to the specified location? What's the fallback strategy? I would recommend a fallback to the existing behaviour, that is to write it to $PWD as to not disrupt support organizations.

YaSuenag commented:
Thanks @mgronlun !
I pushed your change to this PR.

YaSuenag commented:
/integrate

dholmes-ora commented:
This change seems to have caused a number of JFR test failures:
https://bugs.openjdk.java.net/browse/JDK-8275373
https://bugs.openjdk.java.net/browse/JDK-8275375
